feat(rust): define the rust adbc api

wjones127 opened the pull request:
This PR introduces a Rust API for ADBC. This is meant to be parallel to the API in other languages:

 * C: https://github.com/apache/arrow-adbc/blob/main/adbc.h
 * Go: https://github.com/apache/arrow-adbc/blob/main/go/adbc/adbc.go
 * Java: ([Database](https://github.com/apache/arrow-adbc/blob/main/java/core/src/main/java/org/apache/arrow/adbc/core/AdbcDatabase.java), [Connection](https://github.com/apache/arrow-adbc/blob/main/java/core/src/main/java/org/apache/arrow/adbc/core/AdbcConnection.java), [Statement](https://github.com/apache/arrow-adbc/blob/main/java/core/src/main/java/org/apache/arrow/adbc/core/AdbcStatement.java))

It is similar to the others, except for two methods on `AdbcConnection`: `get_info()` and `get_objects()` return Rust data structures instead of Arrow record batch readers. This is to make the API more ergonomic. Both are designed to be convertible from the C API. The info values from `get_info()` are converted from a `UnionArray` to a Rust enum. The `get_objects()` return value is an associated type implementing a trait that defines an easy to use API; the batches returned by C drivers will be wrapped in a struct that provides a zero-copy view ([prototype implementation](https://github.com/apache/arrow-adbc/blob/75ba2cef9080b19f4d242d0a03c17db304609d85/rust/src/driver_manager.rs#L970)).

wjones127 commented:
> It may be useful to return Arrow objects and expose the converter helper? I deemed that a "battle for another day" when working on the R bindings but I suppose you've already written it.

Yes, that's certainly the alternative. I was thinking that the interface would be the Rust types, and there would be a converter to/from the Arrow types for C FFI, but going the other way works to. My only thought is if most of the time people will prefer working with the Rust types, then why not use those as the primary interface?

> I found it a little difficult to picture what using this API might look like while reading the source

Yeah the only reason I don't have examples is I already have them in PRs and I'm running a bit low on steam. But I can point to some of those in the other branches.

Using `get_info()`:

https://github.com/apache/arrow-adbc/blob/75ba2cef9080b19f4d242d0a03c17db304609d85/rust/tests/test_driver_manager.rs#L65-L87

Using `get_objects()`:

https://github.com/apache/arrow-adbc/blob/75ba2cef9080b19f4d242d0a03c17db304609d85/rust/tests/test_driver_manager.rs#L110-L144

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    /// for ADBC usage.  Drivers/vendors will ignore requests for
+    /// unrecognized codes (the row will be omitted from the result).
+    /// Known codes are provided in [info::codes].
+    fn get_info(&self, info_codes: Option<&[u32]>) -> Result<Vec<(u32, InfoData)>, AdbcError>;
```
How about making different functions for this (ignore the fn names):
```rust
fn get_info(&self, info_code: u32) -> Result<(u32, InfoData)>, AdbcError>;
fn get_infos(&self, info_codes: Option<&[u32]>) -> Result<Vec<(u32, InfoData)>, AdbcError>>;
```

wjones127 replied:
Added a `get_single_info` function.

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    fn get_table_types(&self) -> Result<Vec<String>, AdbcError>;
+
+    /// Read part of a partitioned result set.
+    fn read_partition(&self, partition: &[u8]) -> Result<Box<dyn RecordBatchReader>, AdbcError>;
```
Should we use an associated type for the return type?

wjones127 replied:
The driver manager will always return a boxed one, since it might be imported from another language. But if arrow-rs had `impl RecordBatchReader for Box<dyn RecordBatchReader>` then I think we could make this an associated type. The question is, how beneficial would that be? IIUC, it just lets us avoid dynamic dispatch in some cases. I'd be in favor of keeping this idea in mind, but waiting until we have a driver that wants to provide it.

wjones127 replied:
Okay I'm realizing that providing a concrete associated type allows more optimization, so I'll look at doing this as part of a future update once the required trait is implemented upstream.

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+}
+
+/// Depth parameter for GetObjects method.
+#[derive(Debug)]
```
```suggestion
#[derive(Debug, Copy, Clone)]
```

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    fn set_sql_query(&mut self, query: &str) -> Result<(), AdbcError>;
+
+    /// Set the Substrait plan to execute.
+    fn set_substrait_plan(&mut self, plan: &[u8]) -> Result<(), AdbcError>;
```
How about using the `Plan` type provided by the `substrait` crate here?

wjones127 replied:
I'm open to that! Do you think it makes sense to put it behind a crate feature guard, or is it a dependency most won't mind having regardless?

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    /// the corresponding field will be NA (NullType).
+    ///
+    /// This should return an error if this was called before [AdbcStatement::prepare].
+    fn get_param_schema(&mut self) -> Result<Schema, AdbcError>;
```
Does this need `&mut self`?

wjones127 replied:
```suggestion
    fn get_param_schema(&self) -> Result<Schema, AdbcError>;
```

It shouldn't, but the C API uses a non-const pointer here, so if we change to `&self` the driver manager will just have to assume this is safe. ([See discussion](https://github.com/apache/arrow-adbc/pull/417#discussion_r1097581799)). We could regard this quirk of the C API as just "C programmers being C programmers" and consider it a `const` pointer for Rust, or be safety-paranoid and keep as `&mut self`. It's unclear to me if this will be a big impact, as I expect `AdbcStatement` to only be used in a context where someone has a mutable reference to it.

mbrobbel replied:
Hm, I see. Maybe add a comment about this?

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    fn bind_data(&mut self, batch: RecordBatch) -> Result<(), AdbcError>;
+
+    /// Bind Arrow data, either for bulk inserts or prepared statements.
+    fn bind_stream(&mut self, stream: Box<dyn RecordBatchReader>) -> Result<(), AdbcError>;
```
Should the stream be a generic type that implements `RecordBatchReader` instead?

wjones127 replied:
I'm not sure what the benefit would be. The implementor will need to store this RBR somewhere until the query is executed, and it wouldn't make sense to make the Statement or whatever else generic over the RBR type. For the driver manager, this will always be put in a Box for export over C stream interface.

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+        Self: 'a;
+
+    /// List all catalogs in the result set.
+    fn catalogs<'a>(&'a self) -> Box<dyn Iterator<Item = Self::CatalogEntryType<'a>> + 'a>;
```
Should the return type be an associated type?

wjones127 replied:
Yeah this seems like a place where avoiding dynamic dispatch seems worthwhile. :thumbsup:

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+    ///
+    /// Databases that have no notion of catalogs will have one with None for a name.
+    /// This is case sensitive.
+    fn get_catalog<'a>(&'a self, name: Option<&str>) -> Option<Self::CatalogEntryType<'a>> {
```
You can use lifetime elision with `'_`:
```suggestion
    fn get_catalog(&self, name: Option<&str>) -> Option<Self::CatalogEntryType<'_>> {
```

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+        Self: 'a;
+
+    /// List all catalogs in the result set.
+    fn catalogs<'a>(&'a self) -> Box<dyn Iterator<Item = Self::CatalogEntryType<'a>> + 'a>;
```
You can use lifetime elision with `'_`:
```suggestion
    fn catalogs(&self) -> Box<dyn Iterator<Item = Self::CatalogEntryType<'_>> + '_>;
```

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+    fn name(&self) -> Option<&'a str>;
+
+    /// List all schemas in this catalog that are in the result set.
+    fn schemas(&self) -> Box<dyn Iterator<Item = Self::SchemaEntryType> + 'a>;
```
Should this be an associated type?

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+    fn table_type(&self) -> &'a str;
+
+    /// List all the columns in the table.
+    fn columns(&self) -> Box<dyn Iterator<Item = ColumnSchemaRef<'a>> + 'a>;
```
Same question: associated type?

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+    /// Get a column by name.
+    ///
+    /// This is case sensitive. Will return None if the column is not found.
+    fn get_column_by_name(&self, name: &str) -> Option<ColumnSchemaRef<'a>> {
```
```suggestion
    fn column_by_name(&self, name: &str) -> Option<ColumnSchemaRef<'a>> {
```

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    /// List all the constraints on the table.
+    fn constraints(&self) -> Box<dyn Iterator<Item = TableConstraintRef<'a>> + 'a>;
```
Associated type?

mbrobbel commented on rust/src/objects.rs:
```diff
@@ -0,0 +1,483 @@
+    pub constraint_type: TableConstraintTypeRef<'a>,
+}
+
+pub enum TableConstraintTypeRef<'a> {
```
This item is missing a doc comment.

wjones127 commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+
+    /// Execute a query that doesn't have a result set.
+    ///
+    /// Will return the number of rows affected, or -1 if unknown or unsupported.
```
Yes. Maybe I should say something like that.

```suggestion
    /// Will return the number of rows affected. If the affected row count is 
    /// unknown or unsupported by the database, will return `Ok(-1)`.
```

wjones127 commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+/// `result` may be None if there is no meaningful result.
+/// `row_affected` may be -1 if not applicable or if it is not supported.
+pub struct StatementResult {
+    pub result: Option<Box<dyn RecordBatchReader>>,
```
I don't think so. This might be a RBR imported from another language (C/C++/Go) over the C stream interface.

wjones127 commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    ///
+    /// `options` provided will configure the connection, including the isolation
+    /// level. See standard options in [options].
+    fn connect<K, V>(
```
I hadn't at first, since I was mostly thinking in terms of binding to the C API. But I do like the idea for the Rust API, and the bindings to C can spawn a blocking thread to wait.

wjones127 commented on rust/src/info.rs:
```diff
@@ -0,0 +1,287 @@
+pub fn import_info_data(
+    reader: Box<dyn RecordBatchReader>,
+) -> Result<Vec<(u32, InfoData)>, ArrowError> {
+    let batches = reader.collect::<Result<Vec<RecordBatch>, ArrowError>>()?;
+
+    Ok(batches
+        .iter()
+        .flat_map(|batch| {
```
That's an issue, but also the lifetimes of the arrays are a bit messy. I tried this out, but couldn't find an easy solution. I don't think it's worth optimizing much since it's going to be handling very few values.

wjones127 commented on rust/src/info.rs:
```diff
@@ -0,0 +1,287 @@
+            true,
+        ),
+    ])
+}
```
Done.

wjones127 commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    /// Set an option on the database.
+    ///
+    /// Some databases may not allow setting options after it has been initialized.
+    fn set_option(&self, key: &str, value: &str) -> Result<(), AdbcError>;
```
I ended up going with `AsRef<str>`.

wjones127 commented on rust/src/info.rs:
```diff
@@ -0,0 +1,287 @@
+use std::{borrow::Cow, collections::HashMap, sync::Arc};
+
+/// Contains known info codes defined by ADBC.
+pub mod codes {
```
Ended up using `num_enum` to handle this. I like how it turned out 👍

wjones127 commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    /// for ADBC usage.  Drivers/vendors will ignore requests for
+    /// unrecognized codes (the row will be omitted from the result).
+    /// Known codes are provided in [info::codes].
+    fn get_info(&self, info_codes: Option<&[u32]>) -> Result<Vec<(u32, InfoData)>, AdbcError>;
```
Good point. Switched to a HashMap.

wjones127 commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+/// setting `"adbc.connection.autocommit"` to `"false"` (using
+/// [AdbcConnection::set_option]). Turning off autocommit allows customizing
+/// the isolation level. Read more in [adbc.h](https://github.com/apache/arrow-adbc/blob/main/adbc.h).
+pub trait AdbcConnection {
```
Just did this. I made most of the functions that seemed appropriate async, but would welcome input from others if I missed any.

lidavidm replied:
Hmm, how does the community handle having both async and non-async? I guess async is always preferred nowadays?

wjones127 replied:
At least the parts of the Arrow community I've seen, async is always preferred.

wjones127 commented:
Thanks for the review @mbrobbel. Lots of useful pointers. I think I've addressed most of your feedback. LMK your thoughts on the substrait dependency as an optional feature or not.

ConnieLerchHdg commented:
Hey there! So, this PR is pretty exciting - it introduces a Rust API for ADBC that's meant to be just as good as the API in other languages like C, Go, and Java.

The new Rust API is pretty similar to the others, but there are a couple of key differences. For instance, two methods on AdbcConnection - get_info() and get_objects() - return Rust data structures instead of Arrow record batch readers. This makes the API more ergonomic and easier to work with.

Both get_info() and get_objects() are designed to be convertible from the C API, so you don't have to worry about compatibility issues. Plus, the info values from get_info() are converted from a UnionArray to a Rust enum, which makes it even easier to work with.

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// Field Name       | Field Type
+    /// -----------------|--------------
+    /// `table_type`     | `utf8 not null`
+    fn get_table_types(&self) -> Result<Vec<String>, AdbcError>;
```
I think this should also be `async` (and we can drop the `get_` prefix):
```suggestion
    async fn table_types(&self) -> Result<Vec<String>, AdbcError>;
```

wjones127 replied:
I think dropping the `get` prefix makes sense for "getters": functions that return data that is contained within the data structure. But for functions that are going to actually pull data from a database, I think it's appropriate to have the prefix. So I'd like to keep them for these methods on `AdbcConnection`.

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// Get the Arrow schema of a table.
+    ///
+    /// `catalog` or `db_schema` may be `None` when not applicable.
+    async fn get_table_schema(
```
We could drop the `get_` prefix.
```suggestion
    async fn table_schema(
```

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// or more characters, or `"_"` to match exactly one character.  (See
+    /// the documentation of DatabaseMetaData in JDBC or "Pattern Value
+    /// Arguments" in the ODBC documentation.)
+    async fn get_objects(
```
We could drop the `get_` prefix.
```suggestion
    async fn objects(
```

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// Get a single data base metadata. See [AdbcConnection::get_info()].
+    ///
+    /// Will return `None` if the code is not recognized.
+    async fn get_single_info(&self, info_code: InfoCode) -> Result<Option<InfoData>, AdbcError> {
```
We could drop the `get_` prefix:
```suggestion
    async fn single_info(&self, info_code: InfoCode) -> Result<Option<InfoData>, AdbcError> {
```

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// for ADBC usage.  Drivers/vendors will ignore requests for
+    /// unrecognized codes (the row will be omitted from the result).
+    /// Known codes are provided in [info::InfoCode].
+    async fn get_info(
```
We could drop the `get_` prefix.
```suggestion
    async fn info(
```

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// the corresponding field will be NA (NullType).
+    ///
+    /// This should return an error if this was called before [AdbcStatement::prepare].
+    fn get_param_schema(&mut self) -> Result<Schema, AdbcError>;
```
I think this should be `async` and we could drop the `get_` prefix.
```suggestion
    async fn param_schema(&mut self) -> Result<Schema, AdbcError>;
```

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// Turn this statement into a prepared statement to be executed multiple time.
+    ///
+    /// This should return an error if called before [AdbcStatement::set_sql_query].
+    async fn prepare(&mut self) -> Result<(), AdbcError>;
```
How about using a different type for a prepared statement?

wjones127 replied:
I think this isn't a bad idea, but I want to revisit this once I have the driver manager implemented to see if this is possible. Since other APIs don't differentiate these types, it might be hard to do so in Rust.

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+    /// Set the SQL query to execute.
+    fn set_sql_query(&mut self, query: &str) -> Result<(), AdbcError>;
+
+    /// Set the Substrait plan to execute.
+    fn set_substrait_plan(&mut self, plan: substrait::proto::Plan) -> Result<(), AdbcError>;
```
My suggestion would be to make this optional and this method to a subtrait.

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -0,0 +1,408 @@
+/// For use with [crate::AdbcDatabase::set_option],
+/// [crate::AdbcConnection::set_option],
+/// and [crate::AdbcStatement::set_option].
+pub mod options {
```
Should we also create some enums here for the different options sets and then make the option related methods use with these types?

mbrobbel commented on rust/src/lib.rs:
```diff
@@ -261,6 +261,7 @@ pub trait AdbcStatement {
     /// Set the SQL query to execute.
     fn set_sql_query(&mut self, query: &str) -> Result<(), AdbcError>;
 
+    #[cfg(substrait)]
```
Maybe it's better to move this method to a subtrait?
```rust
#[cfg(substrait)]
pub trait AdbcConnectionSubstrait: AdbcConnection {
    /// Set the Substrait plan to execute.
    fn set_substrait_plan(&mut self, plan: substrait::proto::Plan) -> Result<(), AdbcError>;
}
```

wjones127 commented:
I think I've addressed all feedback. Also have new CI successfully run here:

https://github.com/wjones127/arrow-adbc/actions/runs/4522491222

lidavidm commented:
@wjones127 what's the status of this? I can take a glance through and I think we should go ahead and merge it?

wjones127 commented:
I think it's ready to merge, although I don't have time at the moment for any follow up. So creating releases will have to wait.

I have follow up PRs planned for driver manager and utilities for writing C API drivers that are half done.

I think in about 2 months I'll have time to return to this stream to work :)

lidavidm commented:
Great! I'll take a look (and see how much Rust I remember...) and merge after that.

lidavidm commented on rust/src/error.rs:
```diff
@@ -0,0 +1,193 @@
+#[derive(Debug, PartialEq, Copy, Clone)]
+#[repr(u8)]
+pub enum AdbcStatusCode {
+    /// No error.
+    Ok = 0,
```
Does this make sense to have in Rust? It seems `Ok` would be expressed as `Result::Ok` right?

wjones127 replied:
This part is meant to be FFI compatible with the C error codes (probably should have documented that). Although yeah this does provide a redundant state.

Maybe the best solution is to drop this value from the Rust enum, and then handle the conversion of `0` into `Result::Ok` within the C API FFI codebase. 🤔

lidavidm replied:
Yes, keep the Rust definitions appropriate for Rust and do any conversions needed in the FFI part. (Again: the Go and Java definitions make no reference to FFI, though they are still structured as to make FFI reasonable. But Rust should be able to use Rust only without ever having to consider FFI.)

lidavidm commented:
I think let's merge this for now and we can make changes in follow ups. We can hold a vote if/when we want to freeze things (might be easier to wait for 1.1.0 too).
